File: 0383-ransom-note/0383-ransom-note.py
from typing import Dict
class Solution:

    def canConstruct(self, ransomNote: str, magazine: str) -> bool:

        # Create hashmap of a string with its frequency
        def createHashMap(input_string: str) -> Dict[str, int]:
            letter_map = {}
            for t in input_string:
                if letter_map.get(t):
                    letter_map[t] = letter_map[t] + 1
                else:
                    letter_map[t] = 1
            return letter_map

        magazineMap = createHashMap(magazine)

        for t in ransomNote:
            if magazineMap.get(t) is not None and magazineMap.get(t) > 0 :
                magazineMap[t] = magazineMap[t] - 1
            else:
                return False
        
        return True

File: 0383-ransom-note/0383-ransom-note.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {

        // Return of lenght of magazine is smaller than ransomNote
        if(ransomNote.length() > magazine.length())
        {
            return false;
        }
        
        //  Create a new hashmap
        HashMap<Character, Integer> magazineHashMap = new HashMap<>();
        
        // Fill the hashmap with frequency of magazine
        for(int i=0;i<magazine.length();i++){
            char c = magazine.charAt(i);

            int currentCount =  magazineHashMap.getOrDefault(c, 0);
            
            magazineHashMap.put(c, currentCount + 1);
        }

        // Check magazineHashMap for all letter in ransomNote
        for(int j=0; j < ransomNote.length(); j++)
        {
            char t = ransomNote.charAt(j);

            int currentOccurence = magazineHashMap.getOrDefault(t, 0);

            if(currentOccurence==0){
                return false;
            }
            else{
                magazineHashMap.put(t, currentOccurence-1);
            }
        }

        return true;
    }
}